### ISSUE TITLE
make symfony/security-acl optional dependecy

### ISSUE THREAD

**Issue by Tobion** [sonata-project/SonataAdminBundle#7303]
Using Symfony ACL is not recommended for most situations anymore. I also think sonata should not require symfony/security-acl component by default and instead it should be an optional dependency in case people really want acl functionality. This is somehow the case already as sontata would also require symfony/acl-bundle which it does not require. So people need to require certain packages manually to add acl support.

Currently sonata/admin-bundle and doctrine-orm-admin-bundle require symfony/security-acl. Some classes implement interfaces from security-acl like AbstractAdmin implements Symfony\Component\Security\Acl\Model\DomainObjectInterface.
So I think it would be better if sonata runs without acl by default and people wanting acl functionality would just implement acl interfaces in their own classes.

**Comment by VincentLanglet** [comment]
Indeed, we got an issue because someone wasn't using the acl-bundle https://github.com/sonata-project/SonataAdminBundle/issues/7086. So this can be related to https://github.com/sonata-project/SonataAdminBundle/issues/7097.

The Security\Acl namespace is used in
- AbstractAdmin which implements DomainObjectInterface
- GenerateObjectAclCommand
- AdminPermissionMap
- MaskBuilder
- AclSecurityHandler (and interface)
- SecurityExtension
- AdminAclManipulator (and interface)
- AdminObjectAclData
- AdminObjectAclManipulator
- ObjectAclManipulator (and interface)
it also used in the config
```
set('sonata.admin.security.mask.builder.class', MaskBuilder::class)
```

In SonataDoctrineORM, it's just used in the `ObjectAclManipulator`.

What is the way to make this dependency optional ? Do you want to make the PR ?

I assume that the AbstractAdmin won't work if it implements a non-existing interface.

This might be a good idea to move the Acl related class from Util to an Acl folder too @sonata-project/contributors

**Comment by VincentLanglet** [comment]
I created a discussion about it on symfony to get some help
https://github.com/symfony/symfony/discussions/48257

WDYT about this @jordisala1991 ?
How should we update our code to make acl optionnal ?

**Comment by jordisala1991** [comment]
Yes, I think we should make it optional. And if we manage to split it in a separate bundle, even better, never used ACL and when I tried, never got to make it work.

**Comment by VincentLanglet** [comment]
I dunno how we can split since, for example AbstractAdmin implements DomainObjectInterface...

**Comment by jordisala1991** [comment]
Yep, we need to tackle like other things we split from the adminInterface, like breadcrumbs, templates...

Not an easy task tho.

**Comment by eerison** [comment]
I guess this "Adapter" could help in this task
https://github.com/symfony/security-acl/blob/main/Voter/AclVoter.php

edit 1: As discussed here https://github.com/symfony/symfony/discussions/48257, it is recommended use Voters, "maybe" adding support to migrate `acl` to `voter`, and after that deprecate `acl`, it could be removed in next major.

Edit 2: Yeah it isn't soo easy hehehe.

**Comment by dmaicher** [comment]
This issue now actually becomes a bit more pressing since there isn't any new release for `symfony/security-acl` that supports Symfony 8. People need to use `dev-main` for now.

For the `AdstractAdmin` that implements `DomainObjectInterface`: What if we conditionally implement that interface? Something like

```php
if (interface_exists(DomainObjectInterface::class)) {
    /**
     * @internal
     */
    abstract class BaseAbstractAdmin extends AbstractTaggedAdmin implements DomainObjectInterface
    {
        /**
         * @return string
         */
        public function getObjectIdentifier()
        {
            return $this->getCode();
        }
    }
} else {
    /**
     * @internal
     */
    abstract class BaseAbstractAdmin extends AbstractTaggedAdmin
    {
    }
}

abstract class AbstractAdmin extends BaseAbstractAdmin implements AdminInterface, AdminTreeInterface
{
...
}
``` 

I mean technically making `symfony/security-acl` optional will always be a bit of a BC break for people relying on it. They need to explicitly require the package then themselves which is a small BC break. For me personally this would be fine in a minor if we clearly mention this in the release changelog?

So do you think this can even be done in a minor release at all? We can otherwise also look into a new major but this will require some more cleanup work to remove all the deprecated stuff etc.

**Comment by VincentLanglet** [comment]
I'm ok with a small Bc break. Also, I dunno how many people are really using security acl...

**Comment by dmaicher** [comment]
Should we also deprecate everything around ACL? Or shall we postpone this and for now only make it optional?

EDIT: let's first make it optional. Deprecating is some bigger effort. I will look into it.

**Comment by dmaicher** [comment]
Actually I think its not a BC break afterall as people who are using ACL also need to have https://github.com/symfony/acl-bundle installed (which requires `symfony/security-acl`).
